File: internal/app/topic.go
package app

import (
	"github.com/davidmovas/postulator/internal/domain/entities"
	"github.com/davidmovas/postulator/internal/dto"
	"github.com/davidmovas/postulator/pkg/ctx"
	"github.com/davidmovas/postulator/pkg/errors"
)

func (a *App) CreateTopic(topic *dto.Topic) *dto.Response[int] {
	if topic == nil {
		return dtoErr[int](errors.Validation("topic payload is required"))
	}

	entity := &entities.Topic{
		ID:    topic.ID,
		Title: topic.Title,
	}

	id, err := a.topicSvc.CreateTopic(ctx.FastCtx(), entity)
	if err != nil {
		return dtoErr[int](asAppErr(err))
	}

	return &dto.Response[int]{Success: true, Data: id}
}

func (a *App) GetTopic(id int64) *dto.Response[*dto.Topic] {
	t, err := a.topicSvc.GetTopic(ctx.FastCtx(), id)
	if err != nil {
		return dtoErr[*dto.Topic](asAppErr(err))
	}

	return &dto.Response[*dto.Topic]{Success: true, Data: dto.FromTopic(t)}
}

func (a *App) ListTopics(limit, offset int) *dto.PaginatedResponse[*dto.Topic] {
	items, err := a.topicSvc.ListTopics(ctx.FastCtx())
	if err != nil {
		return dtoPagErr[*dto.Topic](asAppErr(err))
	}

	// Normalize pagination parameters
	if limit <= 0 {
		limit = 20
	}
	if offset < 0 {
		offset = 0
	}

	total := len(items)
	start := offset
	if start > total {
		start = total
	}
	end := start + limit
	if end > total {
		end = total
	}

	paged := items[start:end]
	dtos := dto.FromTopics(paged)

	hasMore := end < total

	res := &dto.PaginatedResponse[*dto.Topic]{
		Success: true,
		Items:   dtos,
		Total:   total,
		Limit:   limit,
		Offset:  offset,
		HasMore: hasMore,
	}

	return res
}

func (a *App) UpdateTopic(topic *dto.Topic) *dto.Response[string] {
	if topic == nil {
		return dtoErr[string](errors.Validation("topic payload is required"))
	}

	entity := &entities.Topic{
		ID:    topic.ID,
		Title: topic.Title,
	}

	if err := a.topicSvc.UpdateTopic(ctx.FastCtx(), entity); err != nil {
		return dtoErr[string](asAppErr(err))
	}

	return &dto.Response[string]{Success: true, Data: "updated"}
}

func (a *App) DeleteTopic(id int64) *dto.Response[string] {
	if err := a.topicSvc.DeleteTopic(ctx.FastCtx(), id); err != nil {
		return dtoErr[string](asAppErr(err))
	}

	return &dto.Response[string]{Success: true, Data: "deleted"}
}

func (a *App) AssignTopicToSite(siteID, topicID, categoryID int64, strategy string) *dto.Response[string] {
	strat := entities.TopicStrategy(strategy)
	if strat != entities.StrategyUnique && strat != entities.StrategyVariation {
		return dtoErr[string](errors.Validation("invalid topic strategy"))
	}

	if err := a.topicSvc.AssignToSite(ctx.FastCtx(), siteID, topicID, categoryID, strat); err != nil {
		return dtoErr[string](asAppErr(err))
	}

	return &dto.Response[string]{Success: true, Data: "assigned"}
}

func (a *App) UnassignTopicFromSite(siteID, topicID int64) *dto.Response[string] {
	if err := a.topicSvc.UnassignFromSite(ctx.FastCtx(), siteID, topicID); err != nil {
		return dtoErr[string](asAppErr(err))
	}

	return &dto.Response[string]{Success: true, Data: "unassigned"}
}

func (a *App) GetSiteTopics(siteID int64) *dto.Response[[]*dto.SiteTopic] {
	items, err := a.topicSvc.GetSiteTopics(ctx.FastCtx(), siteID)
	if err != nil {
		return dtoErr[[]*dto.SiteTopic](asAppErr(err))
	}

	return &dto.Response[[]*dto.SiteTopic]{Success: true, Data: dto.FromSiteTopics(items)}
}

func (a *App) GetTopicsBySite(siteID int64) *dto.Response[[]*dto.Topic] {
	items, err := a.topicSvc.GetTopicsBySite(ctx.FastCtx(), siteID)
	if err != nil {
		return dtoErr[[]*dto.Topic](asAppErr(err))
	}

	return &dto.Response[[]*dto.Topic]{Success: true, Data: dto.FromTopics(items)}
}

func (a *App) GetAvailableTopic(siteID int64, strategy string) *dto.Response[*dto.Topic] {
	strat := entities.TopicStrategy(strategy)
	if strat != entities.StrategyUnique && strat != entities.StrategyVariation {
		return dtoErr[*dto.Topic](errors.Validation("invalid topic strategy"))
	}

	t, err := a.topicSvc.GetAvailableTopic(ctx.FastCtx(), siteID, strat)
	if err != nil {
		return dtoErr[*dto.Topic](asAppErr(err))
	}

	return &dto.Response[*dto.Topic]{Success: true, Data: dto.FromTopic(t)}
}

func (a *App) MarkTopicAsUsed(siteID, topicID int64) *dto.Response[string] {
	if err := a.topicSvc.MarkTopicAsUsed(ctx.FastCtx(), siteID, topicID); err != nil {
		return dtoErr[string](asAppErr(err))
	}

	return &dto.Response[string]{Success: true, Data: "marked"}
}

func (a *App) CountUnusedTopics(siteID int64) *dto.Response[int] {
	cnt, err := a.topicSvc.CountUnusedTopics(ctx.FastCtx(), siteID)
	if err != nil {
		return dtoErr[int](asAppErr(err))
	}

	return &dto.Response[int]{Success: true, Data: cnt}
}

File: frontend/src/services/topic.ts
import {
  CountUnusedTopics,
  CreateTopic,
  DeleteTopic,
  GetAvailableTopic,
  GetTopic,
  GetTopicsBySite,
  ImportAndAssignToSite,
  ImportTopics,
  ListTopics,
  MarkTopicAsUsed,
  UpdateTopic,
} from "@/wailsjs/wailsjs/go/app/App";
import { dto } from "@/wailsjs/wailsjs/go/models";
import { unwrapOne, unwrapString, unwrapMany } from "./utils";
import type { TopicStrategy } from "@/constants/topics";

// UI types
export interface Topic {
  id: number;
  title: string;
  createdAt: string;
}

export interface ImportResult {
  totalRead: number;
  totalAdded: number;
  totalSkipped: number;
  added: string[];
  skipped: string[];
  errors: string[];
}

export function mapTopic(x: dto.Topic): Topic {
  return { id: x.id, title: x.title, createdAt: x.createdAt };
}

export function mapImportResult(x: dto.ImportResult): ImportResult {
  return {
    totalRead: x.totalRead,
    totalAdded: x.totalAdded,
    totalSkipped: x.totalSkipped,
    added: x.added,
    skipped: x.skipped,
    errors: x.errors,
  };
}

export interface PaginatedTopics {
  success: boolean;
  items: Topic[];
  total: number;
  limit: number;
  offset: number;
  hasMore: boolean;
  error?: { code: string; message: string; context?: Record<string, any> };
}

export async function listTopics(limit: number, offset: number): Promise<PaginatedTopics> {
  const res = await ListTopics(limit, offset);
  const pr = dto.PaginatedResponse__Postulator_internal_dto_Topic_.createFrom(res as any);
  return {
    success: pr.success,
    items: (pr.items || []).map(mapTopic),
    total: pr.total,
    limit: pr.limit,
    offset: pr.offset,
    hasMore: pr.hasMore,
    error: pr.error ? { code: pr.error.code, message: pr.error.message, context: pr.error.context } : undefined,
  };
}

export async function getTopic(id: number): Promise<Topic> {
  const res = await GetTopic(id);
  return mapTopic(unwrapOne<dto.Topic>(res));
}

export async function createTopic(title: string): Promise<string> {
  const payload = new dto.Topic({ title });
  const res = await CreateTopic(payload);
  return unwrapString(res);
}

export async function updateTopic(id: number, title: string): Promise<string> {
  const payload = new dto.Topic({ id, title });
  const res = await UpdateTopic(payload);
  return unwrapString(res);
}

export async function deleteTopic(id: number): Promise<string> {
  const res = await DeleteTopic(id);
  return unwrapString(res);
}

export async function getTopicsBySite(siteId: number): Promise<Topic[]> {
  const res = await GetTopicsBySite(siteId);
  return unwrapMany<dto.Topic>(res).map(mapTopic);
}

export async function getAvailableTopic(siteId: number, strategy: TopicStrategy): Promise<Topic> {
  const res = await GetAvailableTopic(siteId, strategy);
  return mapTopic(unwrapOne<dto.Topic>(res));
}

export async function importTopics(filePath: string): Promise<ImportResult> {
  const res = await ImportTopics(filePath);
  const dtoRes = unwrapOne<dto.ImportResult>(res as any); // generated type is Response__ImportResult_
  return mapImportResult(dtoRes);
}

export async function importAndAssignToSite(
    filePath: string,
    siteId: number,
    categoryID: number,
    strategy: string
): Promise<ImportResult> {
  const res = await ImportAndAssignToSite(filePath, siteId, categoryID, strategy);
  const dtoRes = unwrapOne<dto.ImportResult>(res as any);
  return mapImportResult(dtoRes);
}

export async function countUnusedTopics(siteId: number): Promise<number> {
  const res = await CountUnusedTopics(siteId);
  return unwrapOne<number>(res as any);
}

export async function markTopicAsUsed(topicId: number, siteId: number): Promise<string> {
  const res = await MarkTopicAsUsed(topicId, siteId);
  return unwrapString(res);
}
